agent/xds/listeners.go

```diff
@@ -542,29 +599,136 @@ func (s *ResourceGenerator) makeIngressGatewayListeners(address string, cfgSnap
 				routePath:       "",
 				httpAuthzFilter: nil,
 			}
-			filter, err := makeListenerFilter(opts)
+
+			// Generate any filter chains needed for services with custom TLS certs
+			// via SDS.
+			sniFilterChains, err := makeSDSOverrideFilterChains(cfgSnap, listenerKey, opts)
 			if err != nil {
 				return nil, err
 			}
 
+			// If there are any sni filter chains, we need a TLS inspector filter!
+			if len(sniFilterChains) > 0 {
+				tlsInspector, err := makeTLSInspectorListenerFilter()
+				if err != nil {
+					return nil, err
+				}
+				listener.ListenerFilters = []*envoy_listener_v3.ListenerFilter{tlsInspector}
+			}
+
+			listener.FilterChains = sniFilterChains
+
+			// See if there are other services that didn't have specific SNI-matching
+			// filter chains. If so add a default filterchain to serve them.
+			if len(sniFilterChains) < len(upstreams) {
+				defaultFilter, err := makeListenerFilter(opts)
+				if err != nil {
+					return nil, err
+				}
+
+				transportSocket, err := makeDownstreamTLSTransportSocket(tlsContext)
+				if err != nil {
+					return nil, err
+				}
+				listener.FilterChains = append(listener.FilterChains,
+					&envoy_listener_v3.FilterChain{
+						Filters: []*envoy_listener_v3.Filter{
+							defaultFilter,
+						},
+						TransportSocket: transportSocket,
+					})
+			}
+
+			resources = append(resources, listener)
+		}
+	}
+
+	return resources, nil
+}
+
+func routeNameForUpstream(l structs.IngressListener, s structs.IngressService) (string, error) {
+	key := proxycfg.IngressListenerKeyFromListener(l)
+
+	// If the upsteam service doesn't have any TLS overrides then it can just use
+	// the combined filterchain with all the merged routes.
+	if !ingressServiceHasSDSOverrides(s) {
+		return key.RouteName(), nil
+	}
+
+	// Return a specific route for this service as it needs a custom FilterChain
+	// to serve it's custom cert so we should attach it's routes to a separate
+	// Route too.
+	return fmt.Sprintf("%s_%s", key.RouteName(), s.ToServiceName().ToServiceID().String()), nil
```

Comment:
TODO: 
 - [ ] This causes conflicts in Ent tests because it means that the default ns service `web` has a route named `web` in OSS but `default/default/web` in Ent. causing all golden files to conflict. This probably needs to follow the same logic `structs.Upstream.Identifier` which omits default values to keep the identifiers consistent between OSS and Ent. Shame we can't easily re-use that method from here with the types available.

Reply:
Fixed in later commit.